doc/manual/contracts.md

```diff
@@ -1198,3 +1202,138 @@ but `std.contract.all_of [Number, String]` is void. Because `fun x => x` can be
+instead of just the problematic field:
+
+```nickel #repl
+> let Schema = std.contract.from_validator (fun val =>
+  if !std.is_record val then
+    'Error { message = "expected a record" }
+  else if !std.record.has_field "server_port" val then
+    'Error { message = "expected a `server_port` field" }
+  else if !std.is_number val.server_port then
+    'Error { message = "expected `server_port` to be a number" }
+  else
+    'Ok
+  )
+  in
+  {
+    server_port = "8080"
+  } | Schema
+error: contract broken by a value
+       expected `server_port` to be a number
+   ┌─ <repl-input-49>:12:3
+   │
+12 │ ╭   {
+13 │ │     server_port = "8080"
+14 │ │   } | Schema
+   │ │       ------ expected type
+   │ ╰───^ applied to this expression
+```
+
+To provide better blame locations for immediate contract failures, the error data
+returned by a custom contract supports an optional `blame_location` field, that
+you can fill out to specify the location that gets blamed. The easiest and most
+common way to take advantage of this is to use `std.contract.check` for checking
+inner values: `std.contract.check` automatically fills out `blame_location` on
+error, and so you will get the more specific error by default.
+
+```nickel #repl
+> let Schema = std.contract.custom (fun label val =>
+  if !std.is_record val then
+    'Error { message = "expected a record" }
+  else if !std.record.has_field "server_port" val then
+    'Error { message = "expected a `server_port` field" }
+  # When this std.contract.check fails, it will return an 'Error { .. }
+  # variant with the `blame_location` field filled out, pointing at
+  # the inner error location.
+  else std.contract.check Number label val.server_port
+  )
+  in
+  {
+    server_port = "8080"
+  } | Schema
+error: contract broken by a value
+   ┌─ <repl-input-50>:13:19
+   │
+13 │     server_port = "8080"
+   │                   ------ evaluated to this expression
+[...]
+```
+
+Most of the time the automatic propagation of `blame_location` should do
+what you want, as long as you use `std.contract.check` for checking
+inner values. For specific purposes, you can manipulate the blame location
+programmatically. For example, here is a contract for a record with two
+small numbers in it. If both numbers are too big, it blames whichever is bigger.
+
+```nickel #repl
+> let SmallNumber = std.contract.from_predicate (fun x => (std.is_number x) && x < 100)
+  in
+  let TwoNumbers =
+    std.contract.custom (fun label value =>
+      std.contract.check { a, b } label value
+      |> match {
+        'Error e => 'Error e,
+        'Ok { a, b } =>
+          [std.contract.check SmallNumber label a, std.contract.check SmallNumber label b]
+          |> match {
+            ['Ok x, 'Ok y] => 'Ok { a = x, b = y },
+            ['Ok _, 'Error e] => 'Error e,
+            ['Error e, 'Ok _] => 'Error e,
+            ['Error { blame_location = a_loc }, 'Error { blame_location = b_loc }] =>
```

Comment:
In this case it works because it happens that `SmallNumber` is a predicate, but I suppose it's more general to expect that the error data might have more than the label.

```suggestion
            ['Error { blame_location = a_loc, ..}, 'Error { blame_location = b_loc, ..}] =>
```